File: solutions/__140_tree_right_side_view.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rightSideView(root *TreeNode) []int {
	view := []int{}
	if root == nil {
		return view
	}
	layer := []*TreeNode{root}
	n := len(layer)
	for n != 0 {
		view = append(view, layer[n-1].Val) // 这一层的最后节点
		for i := 0; i < n; i++ {
			node := layer[0]
			layer = layer[1:]
			// layer中的node永不为nil
			if node.Left != nil {
				layer = append(layer, node.Left)
			}
			if node.Right != nil {
				layer = append(layer, node.Right)
			}
		}
		n = len(layer)
	}
	return view
}

File: solutions/__140_tree_right_side_view.cpp
/*
 *      Author: peanutzhen
 *      Created time: 2021/8/23 18:11:04
 */

#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

void dfs(vector<int> &ans, TreeNode *r, int depth) {
    if (r == nullptr)
        return;
    if (depth == (int) ans.size())
        ans.push_back(r->val);
    dfs(ans, r->right, depth + 1);
    dfs(ans, r->left, depth + 1);
}

vector<int> rightSideView(TreeNode *root) {
    vector<int> ans;
    dfs(ans, root, 0);
    return ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);


    return 0;
}

